### PR TITLE
Makes GWC compliant with CITE expected exceptions

### PR DESCRIPTION
**Pull request by nmco**
Associated issues: 
- https://github.com/GeoWebCache/geowebcache/issues/570
- https://github.com/GeoWebCache/geowebcache/issues/571
- https://github.com/GeoWebCache/geowebcache/issues/573
- https://github.com/GeoWebCache/geowebcache/issues/574

Note that this PR also adds a CITE compliance mode to GeoWebCache, this mode can be activated when using standalone GWC or when using it integrated with GeoServer, related issue is #570.

### CODE REVIEW

**Comment by nmco** [comment]
Failure seems to be unrelated to this PR:
```
[INFO] Web module ......................................... FAILURE [  5.088 s]
```

Locally I can't reproduce this error.

**Comment by bencaradocdavies** [comment]
@nmco I often see this intermittent failure and reported it as https://github.com/GeoWebCache/geowebcache/issues/562. I lack permissions to restart your Travis build or close and reopen your pull request, but either of these options will likely fix your build.

**Comment by nmco** [comment]
I just restart it, thanks for the thip @bencaradocdavies !

File: geowebcache/core/src/main/java/org/geowebcache/config/GeoWebCacheConfiguration.java [hunk 1]
```diff
@@ -314,4 +314,25 @@ public Boolean getFullWMS() {
...
         this.fullWMS = fullWMS;
     }
+
+    /**
+     * If this method returns NULL CITE strict compliance mode should not be considered for WMTS
+     * service implementation.
+     *
+     * @return may return TRUE, FALSE or NULL
+     */
+    public Boolean isWmtsCiteCompliant() {
```
Comment (aaime) [inline geowebcache/core/src/main/java/org/geowebcache/config/GeoWebCacheConfiguration.java]: Confused about the change of signature... what about keeping the field as Boolean for XStream, "boolean" for the methods, and then do a null check inside the body?

File: geowebcache/core/src/main/java/org/geowebcache/config/XMLConfiguration.java [hunk 2]
```diff
@@ -489,6 +488,8 @@ public static XStream getConfiguredXStreamWithContext(XStream xs, WebApplication
 
         xs.alias("serviceInformation", ServiceInformation.class);
         xs.alias("contactInformation", ContactInformation.class);
+
+        xs.omitField(ServiceInformation.class, "citeCompliant");
```
Comment (aaime) [inline geowebcache/core/src/main/java/org/geowebcache/config/XMLConfiguration.java]: Why omit?

File: geowebcache/wmts/src/main/java/org/geowebcache/service/wmts/WMTSService.java [hunk 3]
```diff
@@ -364,4 +402,103 @@ protected SecurityDispatcher getSecurityDispatcher() {
...
+    }
+
+    /**
+     * Helper method that checks if WMTS implementation should be CITE strictly compliant.
+     *
+     * @return TRUE if GWC main configuration or at least one of the WMTS extensions forces
+     *         CITE compliance
+     */
+    private boolean isCiteCompliant() {
+        // let's if main GWC configuration forces WMTS implementation to be CITE compliant
```
Comment (aaime) [inline geowebcache/wmts/src/main/java/org/geowebcache/service/wmts/WMTSService.java]: Let's see?

File: geowebcache/web/src/main/webapp/WEB-INF/geowebcache-wmtsservice-context.xml [hunk 4]
```diff
@@ -22,5 +22,6 @@
     <constructor-arg ref="gwcURLMangler"/>
     <constructor-arg ref="geowebcacheDispatcher"/>
     <property name="securityDispatcher" ref="gwcSecurityDispatcher"/>
+  <property name="mainConfiguration" ref="gwcXmlConfig"/>
```
Comment (aaime) [inline geowebcache/web/src/main/webapp/WEB-INF/geowebcache-wmtsservice-context.xml]: Indent

File: geowebcache/core/src/main/java/org/geowebcache/config/GeoWebCacheConfiguration.java (hunk 1 above)
Comment (nmco) [reply geowebcache/core/src/main/java/org/geowebcache/config/GeoWebCacheConfiguration.java]: I agree, using ``boolean`` makes the API more clean, thank you !

File: geowebcache/core/src/main/java/org/geowebcache/config/XMLConfiguration.java (hunk 2 above)
Comment (nmco) [reply geowebcache/core/src/main/java/org/geowebcache/config/XMLConfiguration.java]: This one is a bit weird indeed. The ``ServiceInfo`` object is used in two distinct situations:
- storing GWC service information (common to WMTS and WMS)
- allow extension points to contribute more information to a specific service, for GeoServer provides specific information for WMTS service

I made CITE compliance be manageable per service, in this case for WMTS service only:
```
<?xml version="1.0" encoding="utf-8"?>
<gwcConfiguration xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns="http://geowebcache.org/schema/1.13.0"
  xsi:schemaLocation="http://geowebcache.org/schema/1.13.0 http://geowebcache.org/schema/1.13.0/geowebcache.xsd">
  <version>1.13.0</version>
  <backendTimeout>120</backendTimeout>
  <wmtsCiteCompliant>true</wmtsCiteCompliant>
```

But the only way extension points have to contribute that information (CITE compliant mode activate \ deactivate) is through the service info object.

The ideal solution where will be to make possible for each service (WMTS, WMS, etc ...) have their own specific configuration, but this goes beyond the scope of this work.

File: geowebcache/web/src/main/webapp/WEB-INF/geowebcache-wmtsservice-context.xml (hunk 4 above)
Comment (nmco) [reply geowebcache/web/src/main/webapp/WEB-INF/geowebcache-wmtsservice-context.xml]: Ouch ... the file is using a mixture of tabs \ spaces, I'm reformatting everything to spaces,

File: geowebcache/wmts/src/main/java/org/geowebcache/service/wmts/WMTSService.java (hunk 3 above)
Comment (nmco) [reply geowebcache/wmts/src/main/java/org/geowebcache/service/wmts/WMTSService.java]: Fixed, thanks !

File: geowebcache/core/src/main/java/org/geowebcache/config/XMLConfiguration.java (hunk 2 above)
Comment (aaime) [reply geowebcache/core/src/main/java/org/geowebcache/config/XMLConfiguration.java]: I see... yeah, rolling a new set of config files, one per service, seems indeed a lot of work.

**Pull request merged**